reconcile/terraform_repo.py

```diff
@@ -0,0 +1,302 @@
+    secret: RepoSecret
+
+
+class TerraformRepoIntegrationParams(PydanticRunParams):
+    output_dir: Optional[str]
+    validate_git: bool
+
+
+class TerraformRepoIntegration(
+    QontractReconcileIntegration[TerraformRepoIntegrationParams]
+):
+    def __init__(self, params: TerraformRepoIntegrationParams) -> None:
+        super().__init__(params)
+        self.qontract_integration = "terraform_repo"
+        self.qontract_integration_version = make_semver(0, 1, 0)
+        self.qontract_tf_prefix = "qrtfrepo"
+
+    @property
+    def name(self) -> str:
+        return self.qontract_integration.replace("_", "-")
+
+    @defer
+    def run(
+        self,
+        dry_run: bool,
+        defer: Optional[Callable] = None,
+    ) -> None:
+
+        gqlapi = gql.get_api()
+
+        state = init_state(integration=self.name)
+        if defer:
+            defer(state.cleanup)
+
+        desired = self.get_repos(query_func=gqlapi.query)
+        existing = self.get_existing_state(state)
+
+        repo_diff = self.calculate_diff(
+            existing_state=existing, desired_state=desired, dry_run=dry_run, state=state
+        )
+
+        for repo in repo_diff:
+            # format each repo into the input the executor expects
+            repo_output = RepoOutput(
+                dry_run=dry_run,
+                repository=repo.repository,
+                name=repo.name,
+                ref=repo.ref,
+                project_path=repo.project_path,
+                delete=repo.delete or False,
+                secret=RepoSecret(
+                    path=repo.account.automation_token.path,
+                    version=repo.account.automation_token.version,
+                ),
+            )
+
+            if self.params.output_dir:
+                try:
+                    output_filename = f"{self.params.output_dir}/{repo.name}.yaml"
+                    with open(output_filename, "w") as output_file:
+                        yaml.safe_dump(
+                            data=repo_output.dict(),
+                            stream=output_file,
+                            explicit_start=True,
+                        )
+                except FileNotFoundError:
+                    raise ParameterError(f"Unable to write to '{output_filename}'")
+            else:
+                print(yaml.safe_dump(data=repo_output.dict(), explicit_start=True))
+
+    def get_repos(self, query_func: Callable) -> list[TerraformRepoV1]:
+        """Gets a list of terraform repos defined in App Interface
+
+        :param query_func: function which queries GQL server
+        :type query_func: Callable
+        :return: list of Terraform repos or empty list if none are defined in A-I
+        :rtype: list[TerraformRepoV1]
+        """
+        query_results = query(query_func=query_func).repos
+        if query_results:
+            return query_results
+        return []
+
+    def get_existing_state(self, state: State) -> list[TerraformRepoV1]:
```

Comment:
General curiosity - as used, is State a terraform concept or a QR concept?

Reply:
It's the QR state. proceed

Reply:
Reading more below and also looking through the State utility. Confusing nomenclature but makes complete sense to me and its a great implementation strategy IMHO